docs/design/jet/023-jobupload.md

```diff
@@ -0,0 +1,164 @@
+| Simulator or Soak Test PR(s)   | _TBD_                                            |
+
+### Background
+#### Description
+
+- Currently, the job upload can only be performed by hz-cli command. This command requires a JVM on the host machine. The purpose of this feature is to enable non-java clients to be able to upload jobs to Jet server 
+
+
+#### Terminology
+
+| Term                   | Definition                                                              |
+|------------------------|-------------------------------------------------------------------------|
+| Client Binary Protocol | The binary messages used between Hazelcast Clients and Hazelcast Server |
+
+#### Actors and Scenarios
+
+Any language that implements Client Binary Protocol
+
+### Functional Design
+#### Summary of Functionality
+
+Provide a list of functions user(s) can perform.
+- Upload a jar file and pass the same parameters that can be used with hz-cli
+
+#### Additional Functional Design Topics
+
+None
+
+##### Notes/Questions/Issues
+
+- The new functionality requires that all resources accessed by the uploaded job to be available on the server. For example a text file that is used to populate an IMap needs to be available on the server or within the uploaded jar. Because a server side job can not access client side resources.  
+- The client should send multi parts in a sequential manner. Out of order messages are not handled since it would require more resources on the server.
+  
+
+### User Interaction
+#### API design and/or Prototypes
+
+```java
+HazelcastInstance client = ...;
+JetService jetService = client.getJet();
+List<String> jobParameters = emptyList();
+
+String job1 = "job1";
+//If there is an error, throws JetException
+jetService.submitJobJar(getJarPath(),
+        null,
+        job1,
+        null,
+        jobParameters);
+```
+
+#### Client Related Changes
+A new method has been added to JetService interface. Any client that implements this interface, has to implement this method
+```java
+void submitJobJar(@Nonnull Path jarPath, String snapshotName, String jobName, String mainClass, List<String> jobParameters);
+```
+
+### Technical Design
+
+The client protocol needs to support job uploading. So that non-java clients can also upload jet jobs.
+For this purpose two new messages are added to client protocol. These messages should be sent to the same member in a cluster since they are logically related
+
+1. uploadJobMetaData
+2. uploadJobMultipart
+
+**1. uploadJobMetaData Message**
+
+The upload process starts with uploadJobMetaData. This message contains
+
+| Term          | Type        | Definition                                                |
+|---------------|-------------|-----------------------------------------------------------|
+| sessionId     | UUID        | The UUID. This field associates all messages in a session |
+| filename      | String      | Name of the jar file without extension                    |
```

Comment:
Does the filename matter on the member side? Is it possible to remove that field altogether? The member can then just use a temporary file to save the Jar.

Reply:
@frant-hartm asked the filename to be added to protocol to make debugging easier.  Quoting him
`well we could use the job name to create the temporary folder, but I think that using the same jar name as on the client would make most sense
most of the time you won't need it, but then when we (or customer) debug some issue they would see:
runjob11545468063021496492.jar
if they don't use job name - they have no idea where the job comes from,
or
BasketCountsJob11545468063021496492.jar
which is admittedly better,
but if they submit e.g. my-job-1.0.jar and they also see the same jar on the server it doesn't leave room for any confusion`

Reply:
In that case how about sending the filename as is, without removing the extension?

Is the jar saved as the filename in this field in the server side? If it is, is it a problem if the same user or other users try to upload a file with the same name in parallel?

Reply:
S3 upload API seems to have checksums for parts too: https://awscli.amazonaws.com/v2/documentation/api/latest/reference/s3api/upload-part.html
The relevant part of the doc:
```
upload-part
[--body <value>]
--bucket <value>
[--content-length <value>]
[--content-md5 <value>]
[--checksum-algorithm <value>]
[--checksum-crc32 <value>]
[--checksum-crc32-c <value>]
[--checksum-sha1 <value>]
[--checksum-sha256 <value>]
--key <value>
--part-number <value>
```

Reply:
We are getting the filename without extension and creating a temporary file with a random value after it. foo.jar becomes fooXYZ.jar . You can upload the same jar in parallel because file names do not conflict